dynamic-check/src/objects/frame.rs

```diff
@@ -0,0 +1,1198 @@
+        self.parse_circuit_member_access(type_, identifier, span)
+    }
+
+    ///
+    /// Returns the type of the accessed circuit member.
+    ///
+    fn parse_circuit_member_access(
+        &mut self,
+        type_: Type,
+        identifier: &Identifier,
+        span: &Span,
+    ) -> Result<Type, FrameError> {
+        // Check that type is a circuit type.
+        let circuit_type = self.parse_circuit_name(type_, span)?;
+
+        // Look for member with matching name.
+        Ok(circuit_type.member_type(&identifier)?)
+    }
+
+    ///
+    /// Returns the type returned by calling the static circuit function.
+    ///
+    fn parse_static_circuit_function_access(
+        &mut self,
+        expression: &Expression,
+        identifier: &Identifier,
+        span: &Span,
+    ) -> Result<Type, FrameError> {
+        // Parse the circuit name.
+        let type_ = self.parse_expression(expression)?;
+
+        self.parse_circuit_member_access(type_, identifier, span)
+    }
+
+    ///
+    /// Returns a `CircuitType` given a circuit expression.
+    ///
+    fn parse_circuit_name(&mut self, type_: Type, span: &Span) -> Result<&CircuitType, FrameError> {
+        // Check that type is a circuit type.
+        match type_ {
+            Type::Circuit(identifier) => {
+                // Lookup circuit identifier.
+                self.user_defined_types
+                    .get_circuit_type(&identifier.name)
+                    .ok_or_else(|| FrameError::undefined_circuit(&identifier))
+            }
+            type_ => Err(FrameError::invalid_circuit(type_, span)),
+        }
+    }
+
+    ///
+    /// Returns a `FunctionType` given a function expression.
+    ///
+    fn parse_function_name(&mut self, expression: &Expression, span: &Span) -> Result<FunctionType, FrameError> {
+        // Case 1: Call a function defined in the program file.
+        // Case 2: Call a circuit function.
+        // Case 3: Call a static circuit function.
+        // Return an Error in any other case.
+        match expression {
+            Expression::Identifier(identifier) => self.parse_program_function(identifier, span),
+            Expression::CircuitMemberAccess(expression, identifier, span) => {
+                self.parse_circuit_function(expression, identifier, span)
+            }
+            Expression::CircuitStaticFunctionAccess(expression, identifier, span) => {
+                self.parse_static_circuit_function(expression, identifier, span)
+            }
+            expression => Err(FrameError::invalid_function(expression, span)),
+        }
+    }
+
+    ///
+    /// Returns a `FunctionType` given a function identifier.
+    ///
+    fn parse_program_function(&mut self, identifier: &Identifier, _span: &Span) -> Result<FunctionType, FrameError> {
+        self.user_defined_types
+            .get_function_type(&identifier.name)
+            .map(|function_type| function_type.to_owned())
```

Comment:
```suggestion
            .cloned()
```
I usually prefer `to_owned` for plain references, but `cloned()` is also clear in terms of specifically converting references to owned objects